src/storage/src/hummock/state_store.rs

```diff
@@ -348,8 +389,49 @@ impl HummockStorage {
             .iter()
             .map(|shared_buffer| shared_buffer.get_overlap_data(key_range))
             .collect();
-
-        Ok((shared_buffer_data, read_version.pinned_version))
+        let sync_uncommitted_data: Vec<OrderSortedUncommittedData> = read_version
+            .sync_uncommitted_datas
+            .iter()
+            .map(|sync_uncommitted_data| match sync_uncommitted_data {
+                Syncing(task) => {
+                    let local_data_iter = task
+                        .iter()
+                        .filter(|(_, data)| match data {
+                            UncommittedData::Batch(batch) => range_overlap(
+                                key_range,
+                                batch.start_user_key(),
+                                batch.end_user_key(),
+                            ),
+                            UncommittedData::Sst((_, info)) => filter_single_sst(info, key_range),
+                        })
+                        .map(|((_, order_index), data)| (*order_index, data.clone()));
+
+                    let mut uncommitted_data = BTreeMap::new();
+                    for (order_index, data) in local_data_iter {
+                        uncommitted_data
+                            .entry(order_index)
+                            .or_insert_with(Vec::new)
+                            .push(data);
+                    }
+                    uncommitted_data.into_values().rev().collect()
+                }
+                Synced(ssts) => vec![ssts
+                    .iter()
+                    .filter(|data| match data {
+                        UncommittedData::Batch(_) => {
+                            panic!("sync uncommitted states can't save batch")
+                        }
+                        UncommittedData::Sst((_, info)) => filter_single_sst(info, key_range),
+                    })
+                    .cloned()
+                    .collect()],
+            })
```

Comment:
Similar to what we did in shared buffer. Please create a `get_overlap_data` method in `SyncUncommittedData` for this logic to make the code more readable.